libvast/src/system/type_registry.cpp

```diff
@@ -86,21 +86,29 @@ caf::error type_registry_state::load_from_disk() {
 }
 
 void type_registry_state::insert(vast::type layout) {
-  auto& cont = data[layout.name()].value;
-  if ([[maybe_unused]] auto [hint, success]
-      = cont.insert(flatten(std::move(layout)));
-      success)
+  // Flatten all incoming types first.
+  auto new_layout = flatten(layout);
+  auto& old_layouts = data[new_layout.name()];
+  // Check whether the new layout is compatible with the latest, i.e., whether
+  // the new layout is a superset of it.
+  if (!old_layouts.empty())
+    if (!is_subset(*old_layouts.begin(), new_layout))
+      VAST_WARNING(self, "detected an incompatible version of",
+                   new_layout.name());
+  // Insert into the existing bucket.
+  auto [hint, success] = old_layouts.insert(std::move(new_layout));
+  if (success)
     VAST_DEBUG(self, "registered", hint->name());
```

Comment:
This can be improved so that `is_subset()` is only called if the new layout was successfully inserted.
```suggestion
  // Insert into the existing bucket.
  auto [hint, success] = old_layouts.insert(std::move(new_layout));
  if (success) {
    if (hint != old_layouts.begin()) {
      if (!is_subset(*old_layouts.begin(), new_layout))
        VAST_WARNING(self, "detected an incompatible version of",
                     new_layout.name());
    }
    VAST_DEBUG(self, "registered", hint->name());
  }
```